### PR TITLE
More documentation

### PR DESCRIPTION
**Pull request by ewjoachim**

### CODE REVIEW

File: docs/discussions.rst [hunk 1]
```diff
@@ -79,11 +78,11 @@ Let's lay down a few words their meaning.
...
     In cabbage, most of the time asynchronoucity is metionned, it's not about
     Python's ``async/await`` features, but about the fact a job is launched
-    in the code, and then the calling code moves on, now waiting for the
-    completion of the job. Because of this, asynchrnous tasks should do
-    something that modifies the world (e.g. compute a costly function **and**
-    store the result somewhere), otherwise the result of the computation will
-    be lost.
+    in the code, and then the calling code moves on, not waiting for the
+    completion of the job. Because of this, asynchronous tasks should have a
+    side effect, and dosomething that modifies the world (e.g. compute a costly
```
Comment (mgu) [inline docs/discussions.rst]:
```suggestion
    side effect, and do something that modifies the world (e.g. compute a costly
```

File: docs/howto.rst [hunk 2]
```diff
@@ -1,23 +1,141 @@
...
+and share resources like a database).
+
+We call it::
+
+    write_alphabet.defer("a")
+    write_alphabet.defer("b")
+    write_alphabet.defer("c")
+    write_alphabet.defer("d")
+
+We could expect the following to be written in `pyramid.txt`::
```
Comment (mgu) [inline docs/howto.rst]:
```suggestion
We could expect the following to be written in `alphabet.txt`::
```

File: docs/howto.rst [hunk 3]
```diff
@@ -1,23 +1,141 @@
...
+    c
+    b
+
+Jobs were taken from the queue in order, but if we have several
+workers, they have been launched in parallel and given their duration
+is random, the final result pretty much is too.
+
+We can solve this problem by using locks::
+
+    job_descritption = write_alphabet.configure(lock="/tmp/alphabet.txt")
```
Comment (mgu) [inline docs/howto.rst]: using a path as lock parameter could be misleading no ? (even if it's the name of the resource, the task might have multiple resources used)

File: docs/howto.rst [hunk 4]
```diff
@@ -1,23 +1,141 @@
...
+    clean.configure(schedule_in={"hours": 1, "minutes": 30}).defer()
+
+The details on the parameters you can use are in the pendulum
+`documentation <https://pendulum.eustace.io/docs/#addition-and-subtraction>`_
+(because we use pendulum under the hood).
+
 Add a task middleware
 ---------------------
 
+There's nothing yet do this, but given you control the decorator
```
Comment (mgu) [inline docs/howto.rst]: I don't understand the first sentence :)

File: docs/howto.rst (hunk 2 above)
Comment (ewjoachim) [reply docs/howto.rst]: Oh yeah right. I started with a different example, and simplified it :D

File: docs/howto.rst [hunk 5]
```diff
@@ -1,23 +1,141 @@
...
-Ensure tasks don't run concurrently and run in order
-----------------------------------------------------
+Ensure jobs don't run concurrently and run in order
+---------------------------------------------------
 
-Launch a task in the future
+Let's imagine we have a task like this::
+
+    @task_manager.task(queue="queue")
+    def write_alphabet(floor):
```
Comment (ewjoachim) [inline docs/howto.rst]: Oh gosh it should NEVER have been called floor anymore ><

Comment (ewjoachim) [inline docs/howto.rst]:
```suggestion
    def write_alphabet(letter):
```

File: docs/howto.rst [hunk 6]
```diff
@@ -1,23 +1,141 @@
...
+        time.sleep(random.random() * 5)
+        with open("/tmp/alphabet.txt", "a") as f:
+            f.write(floor)
```
Comment (ewjoachim) [inline docs/howto.rst]:
```suggestion
            f.write(letter)
```

File: docs/howto.rst [hunk 7]
```diff
@@ -1,23 +1,141 @@
...
+
+We write the letter we recieve to a file after wating for a
+random time (this is a simplified version of a real
+world problem where tasks take a unforseeable amount of time
+and share resources like a database).
+
+We call it::
+
+    write_alphabet.defer("a")
```
Comment (ewjoachim) [inline docs/howto.rst]:
```suggestion
    write_alphabet.defer(letter="a")
```

File: docs/howto.rst [hunk 8]
```diff
@@ -1,23 +1,141 @@
...
+    write_alphabet.defer("b")
```
Comment (ewjoachim) [inline docs/howto.rst]:
```suggestion
    write_alphabet.defer(letter="b")
```

File: docs/howto.rst [hunk 9]
```diff
@@ -1,23 +1,141 @@
...
+    write_alphabet.defer("c")
```
Comment (ewjoachim) [inline docs/howto.rst]:
```suggestion
    write_alphabet.defer(letter="c")
```

File: docs/howto.rst [hunk 10]
```diff
@@ -1,23 +1,141 @@
...
+    write_alphabet.defer("d")
```
Comment (ewjoachim) [inline docs/howto.rst]:
```suggestion
    write_alphabet.defer(letter="d")
```

File: docs/howto.rst (hunk 3 above)
Comment (ewjoachim) [reply docs/howto.rst]: Not sure what you mean... What do you suggest ?

Note: the whole task makes no sense in a distributed context, the file will be split between all the machines that run the workers.

File: docs/howto.rst [hunk 11]
```diff
@@ -1,23 +1,141 @@
...
+
+    job_descritption = write_alphabet.configure(lock="/tmp/alphabet.txt")
+    job_descritption.defer("a")
+    job_descritption.defer("b")
+    job_descritption.defer("c")
+    job_descritption.defer("d")
+
+In this case, our jobs might still be executed by any of the workers,
+but cabbage will not select a job for completion as long as there is
+a job currently processing with the same log. Note that cabbage will
```
Comment (ewjoachim) [inline docs/howto.rst]:
```suggestion
a job currently processing with the same lock. Note that cabbage will
```

File: docs/howto.rst [hunk 12]
```diff
@@ -1,23 +1,141 @@
...
+use postgres to search the jobs table for suitable jobs, meaning that
+even if the database contains a high proportion of locked tasks, it wont
+affect cabbages's capacity to execute the free tasks. Also, identical
```
Comment (ewjoachim) [inline docs/howto.rst]:
```suggestion
affect cabbages's capacity to quickly find the free tasks. Also, identical
```

File: docs/howto.rst [hunk 13]
```diff
@@ -1,23 +1,141 @@
...
+    job_descritption.defer("a")
+    job_descritption.defer("b")
+    job_descritption.defer("c")
+    job_descritption.defer("d")
+
+In this case, our jobs might still be executed by any of the workers,
+but cabbage will not select a job for completion as long as there is
+a job currently processing with the same log. Note that cabbage will
+use postgres to search the jobs table for suitable jobs, meaning that
+even if the database contains a high proportion of locked tasks, it wont
```
Comment (ewjoachim) [inline docs/howto.rst]:
```suggestion
even if the database contains a high proportion of locked tasks, it will barely
```

File: docs/howto.rst [hunk 14]
```diff
@@ -1,23 +1,141 @@
...
+affect cabbages's capacity to execute the free tasks. Also, identical
+jobs will always be started in creation order, so we can be assured our
+tasks will run sequentially and in order.
+
+A good string identifier for the lock is a string identifier of
+the shared resource, UUIDs are well suited for this.
```
Comment (ewjoachim) [inline docs/howto.rst]:
```suggestion
the shared resource, UUIDs are well suited for this.
If multiple resources are implicated, a combination
of their identifiers could be used (there's no hard
limit on the length of a lock string, but stay reasonable)
```

File: docs/howto.rst [hunk 15]
```diff
@@ -1,23 +1,141 @@
...
+
+A task can only take a single lock so there's no dead-lock scenario possible
+where 2 running tasks are waiting one another.
+
+There is no mechanism in place to expire locks yet, but if a task fails
+without the whole Python process to crash, it will free its lock.
```
Comment (ewjoachim) [inline docs/howto.rst]:
```suggestion
without the whole Python process crashing, it will free its lock.
```

File: docs/howto.rst (hunk 3 above)
Comment (mgu) [reply docs/howto.rst]: ok, keep this :)

**Pull request merged**